docs/design-documents/20240430-schema-support.md

```diff
@@ -0,0 +1,394 @@
+      * [Chosen option](#chosen-option)
+    * [Schema definition](#schema-definition)
+    * [Schema service interface](#schema-service-interface)
+      * [Option 1: Stream of commands and responses](#option-1-stream-of-commands-and-responses)
+      * [Option 2: Exposing a gRPC service in Conduit](#option-2-exposing-a-grpc-service-in-conduit)
+      * [Chosen option](#chosen-option-1)
+  * [How are requirements addressed](#how-are-requirements-addressed)
+  * [Summary](#summary)
+  * [Other considerations](#other-considerations)
+<!-- TOC -->
+
+## The problem
+
+A Conduit user would like to start using a destination connector **without**
+manually setting up
+the [target collections](https://conduit.io/docs/introduction/vocabulary).
+
+To do so, the following is required:
+
+- collection name
+- collection metadata
+- schema
+
+This design document focuses on the schema support in Conduit in the context of
+the above.
+
+## Open questions
+
+1. **Q**: Should Conduit allow schemas from other streaming tools?
+
+   Conduit is sometimes used together with other streaming tools, such as Kafka
+   Connect. For example, Kafka Connect is used to stream data to a topic, and
+   then Conduit is used to read the data from the topic and write it into a
+   destination.
+
+   **A**:
+2. **Q**: Should schemas be available in processors?
+
+## Requirements
+
+1. Records **should not** carry the whole schema.
+
+   Reason: If a record would carry the whole schema, that might increase a
+   record's size significantly.
+2. Sources and destinations need to be able to work with multiple schemas.
+   
+   Reason: Multiple collections support.
+3. A schema should be accessible across pipelines and Conduit instances.
+
+   Reason: Pipelines might work together to stream data from A to B, through an
+   intermediary C. For example, PostgreSQL to Kafka, and then Kafka to
+   Snowflake. Also, in the future, pipelines might run in isolation on different
+   instances. Additionally, Conduit may be used so that there's one pipeline per
+   instance. In such cases, we'd need this feature.
+4. It should be possible for a schema to evolve.
+5. A source connector should be able to register a schema.
+6. A destination connector should be able to fetch a specific schema.
+7. A destination connector needs a way to know that a schema changed.
+8. The Connector SDK should provide an API to work with the schemas. 
+9. The Connector SDK should cache the schemas.
+
+   Reason: Standalone connectors and Conduit communicate via gRPC. To avoid the
+   cost of repeatedly fetching or creating the same schema many times (
+   especially over gRPC), schemas should be cached by the SDK.
+
+## Non-goals
+
+- Schema auto-generation: Given that a source connector usually has direct
+  access to a source collection, and also knows most about the source types,
+  schema auto-generation is not a requirement.
+
+## Schema format
```

Comment:
This section describes the types we need. I'll mention our options for the actual schema format (our own, Avro, Protobuf schema, etc.)